Spigot-API-Patches/0021-Add-exception-reporting-event.patch

```diff
@@ -0,0 +1,605 @@
++import org.github.paperspigot.event.ServerExceptionEvent;
++import org.github.paperspigot.exception.ServerPluginMessageException;
+ 
+ /**
+  * Standard implementation to {@link Messenger}
+@@ -428,6 +430,11 @@ public class StandardMessenger implements Messenger {
+             } catch ( Throwable t )
+             {
+                 org.bukkit.Bukkit.getLogger().log( java.util.logging.Level.WARNING, "Could not pass incoming plugin message to " + registration.getPlugin(), t );
++                // Paper start
++                source.getServer().getPluginManager().callEvent(new ServerExceptionEvent(
++                        new ServerPluginMessageException(t, registration.getPlugin(), source, channel, message)
++                ));
++                // Paper end
+             }
+             // Spigot End
+         }
+diff --git a/src/main/java/org/github/paperspigot/event/ServerExceptionEvent.java b/src/main/java/org/github/paperspigot/event/ServerExceptionEvent.java
+new file mode 100644
+index 0000000..6f046cb
+--- /dev/null
++++ b/src/main/java/org/github/paperspigot/event/ServerExceptionEvent.java
+@@ -0,0 +1,36 @@
++package org.github.paperspigot.event;
++
++import org.apache.commons.lang.Validate;
++import org.bukkit.event.Event;
++import org.bukkit.event.HandlerList;
++import org.github.paperspigot.exception.ServerException;
++
++/**
++ * Called whenever an exception is thrown in a recoverable section of the server.
++ */
++public class ServerExceptionEvent extends Event {
++    private static final HandlerList handlers = new HandlerList();
++    private ServerException exception;
++
++    public ServerExceptionEvent (ServerException exception) {
++        Validate.notNull(exception);
++        this.exception = exception;
++    }
++
++    /**
++     * Gets the wrapped exception that was thrown.
++     * @return Exception thrown
++     */
++    public ServerException getException() {
++        return exception;
++    }
++
++    @Override
++    public HandlerList getHandlers() {
++        return handlers;
++    }
++
++    public static HandlerList getHandlerList() {
++        return handlers;
++    }
++}
+diff --git a/src/main/java/org/github/paperspigot/exception/ServerCommandException.java b/src/main/java/org/github/paperspigot/exception/ServerCommandException.java
+new file mode 100644
+index 0000000..39a4521
+--- /dev/null
++++ b/src/main/java/org/github/paperspigot/exception/ServerCommandException.java
+@@ -0,0 +1,68 @@
++package org.github.paperspigot.exception;
++
++import org.apache.commons.lang.Validate;
++import org.bukkit.command.Command;
++import org.bukkit.command.CommandSender;
++
++/**
++ * Thrown when a command throws an exception
++ */
++public class ServerCommandException extends ServerException {
++    public ServerCommandException(String message, Throwable cause, Command command, CommandSender commandSender, String[] arguments) {
++        super(message, cause);
++        Validate.notNull(command);
```

Comment:
If you use Guava's `Preconditions.checkNotNull` method you can do:
`this.command = checkNotNull(command, "command");`

(applies to all cases of this)